### PR TITLE
(GH-15) Documentation added.

### PR DESCRIPTION
**Pull request by stefan-lindegger**

### CODE REVIEW

File: docs/input/docs/getting-started/principles.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/principles.md]: Please add some line breaks, the rendered version is incredibly hard to read.
grafik

File: docs/input/docs/getting-started/whystrategypattern.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+* Support the usage of strategies with no need to implement specific strategy factories
+* Provide a generic strategy provider based on generics
+* Introduce strategy categorization based on generic parameter
+
+Implementation details of BBT.StrategyPattern are:
+
+* Introduces a generic strategy interface
+* The generic parameter represents the criterion for the selection of the specific strategy
+* Works with any IoC framework
+* Supports generic type unbound IoC registrations
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/whystrategypattern.md]: "Generic type unbound IoC registrations" are a feature of the IoC container, not this library.

File: docs/input/docs/getting-started/whystrategypattern.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+---
+Order: 10
+Title: Why does BBT.StrategyPattern exist?
+Description: Introduction to BBT.StrategyPattern, what problems are solved and how it is distinctive from other implementations of strategy pattern.
+---
+
+## Introduce structure to your code by using BBT.StrategyPattern
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/whystrategypattern.md]: Also document the fact that using one library based strategy pattern implementation let's you reuse the strategies across multiple projects and libraries if neccessary.

File: docs/input/docs/getting-started/principles.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+Order: 20
+Title: Core principles
+Description: Core principles of BBT.StrategyPattern.
+---
+
+Goal of this infrastructure is to support the providing of specific strategies achieved by usage of generics.
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/principles.md]:
```suggestion
Goal of this infrastructure is to support the providing of specific strategies achieved by using generics.
```

File: docs/input/docs/getting-started/principles.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+This infrastructure introduces a generic strategy interface managed by generic strategy provider.
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/principles.md]:
```suggestion
This infrastructure introduces a generic strategy interface managed by the generic strategy provider.
The library introduces an `IGenericStrategy<T>` which needs to be implemented for every case a strategy with a given criterion is supporting. The instances of the `IGenericStrategy<T>` are managed by the `IGenericStrategyProvider<out TStrategy, in TCriterion>` which resolves the correct implementation for the given criterion.
```

File: docs/input/docs/getting-started/principles.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+The generic parameter acts as selection criterion and declares the responsibility of the specific strategy.
+The generic strategy interface leads to more structured code.
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/principles.md]: that's a feature, not a core principle

File: docs/input/docs/getting-started/principles.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+Based on its generic type parameter, strategies are categorized and make their responsibility explicit.
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/principles.md]:
```suggestion
Based on their generic type parameter, strategies are categorized and make their responsibility explicit.
```

File: docs/input/docs/getting-started/whystrategypattern.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+BBT.StrategyPattern is an implementation of the strategy design pattern.
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/whystrategypattern.md]: Remove sentence, we had this in the introduction.

File: docs/input/docs/getting-started/fundamentals.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+#### Open/closed principle
+
+When FooEnum is enhanced, with the generic strategy provider, additional strategies can be added without modifying existing classes.
+As soon as the new strategies are registered in the IoC container, the provider is extended.
+
+#### Single responsibility principle
+
+The generic provider has one single responsibility: Providing the requested strategy. How this strategy is instantiated is separated and handled be the strategy locator (i.e. IoC container).
+
+[Strategy pattern (wikipedia)]: https://en.wikipedia.org/wiki/Strategy_pattern
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/fundamentals.md]:
```suggestion
[Strategy pattern (Wikipedia)]: https://en.wikipedia.org/wiki/Strategy_pattern
```

File: docs/input/docs/getting-started/fundamentals.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+[Strategy and open/closed principle]: https://en.wikipedia.org/wiki/Strategy_pattern#Strategy_and_open.2Fclosed_principle
+[Open/close principle (wikipedia)]: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/fundamentals.md]:
```suggestion
[Open/close principle (Wikipedia)]: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle
```

File: docs/input/docs/getting-started/fundamentals.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+[Single responsibility principle (wikipedia)]: https://en.wikipedia.org/wiki/Single_responsibility_principle
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/fundamentals.md]:
```suggestion
[Single responsibility principle (Wikipedia)]: https://en.wikipedia.org/wiki/Single_responsibility_principle
```

File: docs/input/docs/getting-started/fundamentals.md [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+    public void Foo(FooEnum fooEnum)
+    {
+        var barStrategy = this.FooStrategyFactory.GetStrategy(fooEnum);
+        barStrategy.Bar();
+    }
+}
+```
+
+### Improvements of usage of strategy
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/fundamentals.md]:
```suggestion
### Improvements with strategy usage
```

File: docs/input/docs/getting-started/fundamentals.md [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+// Generic type unbound registration of strategy locator.
+// When resolved a locator object of concrete generic type is instantiated.
+kernel.Bind(typeof(IStrategyLocator<>)).To(typeof(NinjectStrategyLocator<>));
+
+// Bindings of the foo strategy implementations.
+kernel.Bind<IFooStrategy>().To<Foo1Strategy>();
+kernel.Bind<IFooStrategy>().To<Foo2Strategy>();
+```
+
+### Improvements of usage of generic strategy and provider
```
Comment (marco-bertschi) [inline docs/input/docs/getting-started/fundamentals.md]:
```suggestion
### Improvements with generic strategy and generic strategy provider usage
```

File: docs/input/docs/getting-started/whystrategypattern.md (hunk 3 above)
Comment (stefan-lindegger) [reply docs/input/docs/getting-started/whystrategypattern.md]: This would be true for every infrastructure?

**Pull request merged**